### PR TITLE
LS: Cancellation support

### LINKED ISSUE #6131: Cancel pending requests when didChange & et all appear.
Then any of the `textDocument/did*` requests come, we need to cancel all operations that are holding snapshots of the `AnalysisDatabase` in order to unlock mutations to it.

This task is spin off #5355

### PR DESCRIPTION
**Pull request by Draggu**
fix #6131

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)

### CODE REVIEW

**Review by piotmag769 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263#-:-O4zDBjv9VKf3dk1jg2y:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Arcticae, @mkaput, and @orizi)

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: all files reviewed, 1 unresolved discussion (waiting on @Arcticae, @Draggu, and @mkaput)

___
*[`crates/cairo-lang-language-server/src/lib.rs` line 152 at r1](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O57W4YEBm7qC4mC3TOI:-O57W4YEBm7qC4mC3TOJ:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/a6a7ad238c609d9aad88cbaac49df848e54bf490/crates/cairo-lang-language-server/src/lib.rs#L152)):*
> ```XML
> 
>     let (service, socket) = Backend::build_service(tricks);
>     Server::new(stdin, stdout, socket).concurrency_level(100).serve(service).await;
> ```





_Suggestion:_
```XML
    Server::new(stdin, stdout, socket).concurrency_level(SOME_CONSTANT_WITH_A_GOOD_NAME).serve(service).await;
```

**Review by mkaput — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: all files reviewed, 6 unresolved discussions (waiting on @Arcticae and @Draggu)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DL3RNCxLucKNLcpfm):*




_Suggestion:_
```
- f774b6c: LS: Cancellation support

  fix #6131

  commit-id:c6b3800a
```
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 120 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DJHe6CODnccs061EG:-O5DJHe6CODnccs061EH:b1l5omr) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L120)):*
> ```XML
> /// [lib]: crate#running-with-customizations
> #[non_exhaustive]
> #[derive(Default, Clone)]
> ```

We should avoid changing `Tricks` at all costs. The whole purpose of this thing is to be super stable, so that we won't bother downstream users.
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 139 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DKne5DhHCQeXU9mAX:-O5DKne5DhHCQeXU9mAY:b9k2nvg) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L139)):*
> ```XML
> /// Number of LSP requests that can be processed concurently.
> /// Higher number than default 4 because cancellation will skip not needed ones, and latest ones can
> /// be processed.
> ```

this sentence is hard to parse: what are `ones` in this context? where does the default come from? what do you mean be `needed`?
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 331 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DKMB96XbAsgGxxeLF:-O5DKMB96XbAsgGxxeLG:b-toxkn0) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L331)):*
> ```XML
>             catch_unwind(AssertUnwindSafe(|| f(&db))).map_err(|err| {
>                 if err.is::<Cancelled>() {
>                     info!("LSP worker thread was cancelled");
> ```

not sure we need `info` level for this. this is probably going to be noisy in long-term, and this is not actionable to the user and self-fixing, isn't it?



_Suggestion:_
```XML
debug!("LSP worker thread was cancelled");
```
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 514 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DK67P7R8kg8j-BX9V:-O5DK67P7R8kg8j-BX9W:b-t396fh) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L514)):*
> ```XML
>             .with_db({
>                 let open_files = open_files.clone();
>                 let tricks = self.tricks.clone();
> ```

why not just construct `new_db` outside `with_db` and move it into it? this way you'd avoid needing to clone `tricks`

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: all files reviewed, 5 unresolved discussions (waiting on @Arcticae and @Draggu)

___
*[`crates/cairo-lang-language-server/src/lib.rs` line 120 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DJHe6CODnccs061EG:-O5I5rLPEz3tzD5lPYxi:b-6z9iqi) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L120)):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

We should avoid changing `Tricks` at all costs. The whole purpose of this thing is to be super stable, so that we won't bother downstream users.
</blockquote></details>

discussed this offline. let's add this `Clone` due to all the issues lack of it brings in this PR

**Review by Draggu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @Arcticae, @mkaput, and @orizi)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DL3RNCxLucKNLcpfm):*
Done.
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 152 at r1](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O57W4YEBm7qC4mC3TOI:-O5IE4PN8KsSj2tstd6V:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/a6a7ad238c609d9aad88cbaac49df848e54bf490/crates/cairo-lang-language-server/src/lib.rs#L152)):*
> ```XML
> 
>     let (service, socket) = Backend::build_service(tricks);
>     Server::new(stdin, stdout, socket).concurrency_level(100).serve(service).await;
> ```

Done.
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 139 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DKne5DhHCQeXU9mAX:-O5IE3rfCb__lC3o25sO:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L139)):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

this sentence is hard to parse: what are `ones` in this context? where does the default come from? what do you mean be `needed`?
</blockquote></details>

Done.
___
*[`crates/cairo-lang-language-server/src/lib.rs` line 331 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6263#-O5DKMB96XbAsgGxxeLF:-O5IE4wc05ttde_Yj8nb:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/f774b6c452cc39ceadadd467b2cb1ec0f82a668a/crates/cairo-lang-language-server/src/lib.rs#L331)):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

not sure we need `info` level for this. this is probably going to be noisy in long-term, and this is not actionable to the user and self-fixing, isn't it?
</blockquote></details>

Done.

**Review by mkaput — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263#-:-O5IFIVgCcY_rwxGdp4u:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @Arcticae and @orizi)

**Review by Arcticae — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263#-:-O5IHm4m9M--6sIRfeGv:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @orizi)

**Review by mkaput — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: all files reviewed, 1 unresolved discussion (waiting on @orizi)

**Review by mkaput — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: all files reviewed, 1 unresolved discussion (waiting on @orizi)

**Review by orizi — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263#-:-O5IKfXrECsYFpJPM7CB:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Draggu)

**Review by mkaput — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Draggu)

**Review by piotmag769 — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6263)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Draggu)
